router/src/lib.rs

```diff
@@ -896,52 +896,99 @@ pub(crate) struct ImageUrl {
+            .0
+            .iter()
+            .map(|chunk| match chunk {
+                ContentChunk::Text(s) => s.clone(),
+                ContentChunk::ImageUrl(s) => format!("![]({})", s),
+            })
+            .collect::<Vec<_>>()
+            .join(" ");
+        serializer.serialize_str(&formatted)
+    }
 }
 
 mod message_content_serde {
     use super::*;
-    use serde::de;
-    use serde::Deserializer;
+    use serde::de::{Deserialize, Deserializer, Error};
     use serde_json::Value;
 
-    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<ContentChunks>, D::Error>
     where
         D: Deserializer<'de>,
     {
         let value = Value::deserialize(deserializer)?;
-        match value {
-            Value::String(s) => Ok(Some(s)),
-            Value::Array(arr) => {
-                let results: Result<Vec<String>, _> = arr
-                    .into_iter()
-                    .map(|v| {
-                        let content: Content =
-                            serde_json::from_value(v).map_err(de::Error::custom)?;
-                        match content.r#type.as_str() {
-                            "text" => Ok(content.text.unwrap_or_default()),
-                            "image_url" => {
-                                if let Some(url) = content.image_url {
-                                    Ok(format!("![]({})", url.url))
-                                } else {
-                                    Ok(String::new())
-                                }
-                            }
-                            _ => Err(de::Error::custom("invalid content type")),
-                        }
-                    })
-                    .collect();
 
-                results.map(|strings| Some(strings.join("")))
-            }
+        match value {
+            Value::String(s) => Ok(Some(vec![s.into()].into())),
+            Value::Array(arr) => arr
+                .into_iter()
+                .map(|v| match v {
+                    Value::String(s) => Ok(ContentChunk::Text(s)),
+                    Value::Object(map) => match map
+                        .get("image_url")
+                        .and_then(|x| x.get("url").and_then(|u| u.as_str()))
+                    {
+                        Some(url) => Ok(ContentChunk::ImageUrl(url.to_string())),
+                        None => map
+                            .get("text")
+                            .and_then(|t| t.as_str())
+                            .map(|text| Ok(ContentChunk::Text(text.to_string())))
+                            .map_or_else(
+                                || Err(Error::custom("Expected a string or an object")),
+                                |x| x,
+                            ),
+                    },
+                    _ => Err(Error::custom("Expected a string or an object")),
+                })
```

Comment:
You shouldn't need this either. Serde should be able to take care of it.

```rust
struct ImageUrl{
    image_url: String
}
```

Reply:
updated and relies on serde to extract the values for both image and text now